devito/core/gpu_openmp.py

```diff
@@ -212,6 +220,82 @@ def _(iet):
         return iet, {}
 
 
+@iet_pass
+def initialize(iet, **kwargs):
+    """
+    Initialize the OpenMP environment.
+    """
+    devicenum = Symbol(name='devicenum')
+
+    @singledispatch
+    def _initialize(iet):
+        comm = None
+
+        for i in iet.parameters:
+            if isinstance(i, MPICommObject):
+                comm = i
+                break
+
+        if comm is not None:
+            rank = Symbol(name='rank')
+            rank_decl = LocalExpression(DummyEq(rank, 0))
+            rank_init = Call('MPI_Comm_rank', [comm, Byref(rank)])
+
+            ngpus = Symbol(name='ngpus')
+            call = Function('omp_get_num_devices')()
+            ngpus_init = LocalExpression(DummyEq(ngpus, call))
+
+            devicenum_init = LocalExpression(DummyEq(devicenum, rank % ngpus))
+
+            body = [rank_decl, rank_init, ngpus_init, devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0+MPI setup'),
+                        body=body,
+                        footer=(c.Comment('End of OpenMP5.0+MPI setup'), c.Line()))
+        else:
+            devicenum_init = LocalExpression(DummyEq(devicenum, 0))
+            body = [devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0 setup'),
+                        body=body,
+                        footer=(c.Comment('End of OpenMP5.0 setup'), c.Line()))
+
+        iet = iet._rebuild(body=(init,) + iet.body)
+
+        return iet
+
+    @_initialize.register(ElementalFunction)
+    @_initialize.register(MPICallable)
+    def _(iet):
+        return iet
+
+    iet = _initialize(iet)
+
+    return iet, {'args': devicenum}
+
+
+@iet_pass
+def mpi_gpu_direct(iet, **kwargs):
+    """
+    Modify MPI Callables to enable multiple GPUs performing GPU-Direct communication.
+    """
+    def _mpi_gpu_direct(iet):
```

Comment:
why do you need this nested function?

since you're not using single dispatch, and since you're calling it just once, I think it's uselss?

Reply:
Agreed. I'll push that together with the next changes.

Reply:
thanks!